fillMessageKeys: allow for SessionCipher creator to specify limit

scottnonnenberg opened the pull request:
People leave their laptops closed for weeks at a time, and [get this error](https://github.com/WhisperSystems/Signal-Desktop/issues/1297), since their other devices are sending messages to it constantly:

"Too many message keys for chain"

And it seems to be really hard to fix once you're in this state. Sync messages no longer show up from the device that got into this state.

liliakai commented:
Do we have a theory on why this seems specific to sync messages?

liliakai commented on src/SessionCipher.js:
```diff
@@ -240,7 +240,10 @@ SessionCipher.prototype = {
   },
   fillMessageKeys: function(chain, counter) {
-      if (Object.keys(chain.messageKeys).length >= 1000) {
+      var ourNumber = textsecure.storage.user.getNumber();
```
textsecure.storage is not part of this repo and we shouldn't depend on it.

scottnonnenberg replied:
How else might a user of this library tell us who the current user is?

liliakai replied:
I don't think this library should care about a "current user" concept, as it generally operates on a per-device level. I would just make the limit an option on SessionCipher, then the application can decide:
```
new SessionCipher(store, address); // limit 1000
new SessionCipher(store, address, isOurNumber ? { fillMessageKeysLimit: 5000 }); // limit 5000
new SessionCipher(store, address, isOurNumber ? { fillMessageKeysLimit: false }); // no limit
```

liliakai reviewed (dismissed):
It'd be best to do this without introducing cross-repo dependencies. This change would create errors for anyone who happens to be using this library outside Signal Desktop.

Maybe we could consider simply raising this limit by default or adding it as a configurable option to SessionCipher's constructor?

scottnonnenberg commented:
This PR has been updated with @liliakai's recommendations. If `options.messageKeysLimit` is provided and falsey, then we don't apply any limit at all. This can be used to set no limit for communications to and from your own devices.

scottnonnenberg committed:
SessionCipher: allow caller to provide fillMessageKeys limit

If options.messageKeysLimit is provided by falsey, then we don't apply
any limit at all. This can be used to set no limit for communications
from your own devices.

Why would you want that? People leave their laptops closed for weeks at
a time and get this error, since their other devices are sending
messages to it constantly:

"Too many message keys for chain"

And it seems to be really hard to fix once you're in this state. Sync
messages no longer show up from the device that got into this state.

FREEBIE

michaelkirk commented:
> Do we have a theory on why this seems specific to sync messages?

My understanding is that this most frequently affects sync messages  because if you have `n` conversations with `k` messages, each of those chains will be only `k` long, but the chain with your synced device will be `n*k`.

Right?

michaelkirk reviewed (approved):
LGTM

My understanding is that if the device is offline for "a long enough time" we'll also see message loss due to the server queue being full. But I think this is a step in the right direction. It seems reasonable that sync queues/chains could get special treatment.

scottnonnenberg commented:
Exactly. You communicate far, far more with your own devices than anyone else.

6531076392 commented:
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform"  
                xmlns:msxsl="urn:schemas-microsoft-com:xslt"
                xmlns:user="http://android.riteshsahu.com">
<xsl:template match="/">
  <html>
	  <head>
		  <style type="text/css">
		  body 
		  {
			font-family:arial,sans-serif;
			color:#000;
			font-size:13px;
			color:#333;
		  }
		  table 
		  {
			font-size:1em;
			margin:0 0 1em;
			border-collapse:collapse;
			border-width:0;
			empty-cells:show;
		  }
		  td,th 
		  {
			border:1px solid #ccc;
			padding:6px 12px;
			text-align:left;
			vertical-align:top;
			background-color:inherit;
		  }
		  th 
		  {
			background-color:#dee8f1;
		  }
		  </style>
	  </head>
	  <body>
	  <h2>Call Logs</h2>
	  <table>
		<tr>
		  <th>Type</th>
		  <th>Number</th>
		  <th>Contact</th>
		  <th>Date</th>
		  <th>Duration (sec)</th>
		</tr>
		<xsl:for-each select="calls/call">
		<tr>
		  <td>
			<xsl:if test="@type = 1">
			Incoming
			</xsl:if>
			<xsl:if test="@type = 2">
			Outgoing
			</xsl:if>
			<xsl:if test="@type = 3">
			Missed
			</xsl:if>
		  </td>
		  <td><xsl:value-of select="@number"/></td>
		  <td><xsl:value-of select="@contact_name"/></td>
		  <td><xsl:value-of select="@date"/><br/><xsl:value-of select="@readable_date"/></td>
		  <td><xsl:value-of select="@duration"/></td>
		</tr>
		</xsl:for-each>
	  </table>
	  </body>
  </html>
</xsl:template>
</xsl:stylesheet>
